### PR TITLE
bug/24-fix-scalar-functions

### LINKED ISSUE #24: Fix broken scalar functions
The following scalar functions seem to be broken:

- [ ] GREATEST
- [ ] LEAST
- [ ] ROUND
- [ ] CONCAT
- [ ] INSTR
- [ ] UNICODECHR
- [ ] UNICODE
- [x] ADD_DAYS
- [x] ADD_HOURS
- [x] ADD_MINUTES
- [x] ADD_MONTHS
- [ ] HOURS_BETWEEN
- [x] ADD_SECONDS
- [x] ADD_WEEKS
- [x] ADD_YEARS
- [ ] SECONDS_BETWEEN
- [ ] MINUTES_BETWEEN
- [ ] SECOND
- [ ] TO_CHAR
- [ ] POSIX_TIME

## Todo for each:

* Remove from `PostgreSQLScalarFunctionsIT#KNOWN_BROKEN` to re-enable the test.
* Fix function

### PR DESCRIPTION
**Pull request by jakobbraun**
Closes #24

### CODE REVIEW

File: src/test/java/com/exasol/adapter/dialects/postgresql/PostgreSQLScalarFunctionsIT.java [hunk 1]
```diff
@@ -35,7 +27,7 @@
 
     @Override
     protected Set<ScalarFunctionCapability> getDialectSpecificExcludes() {
-        return KNOWN_BROKEN;
+        return Set.of();
```
Comment (chiaradiamarcelo) [inline src/test/java/com/exasol/adapter/dialects/postgresql/PostgreSQLScalarFunctionsIT.java:30]:
```suggestion
        return Collections.emptySet();
```

File: src/test/java/com/exasol/adapter/dialects/postgresql/PostgreSQLSqlDialectTest.java [hunk 2]
```diff
@@ -81,17 +80,8 @@ void testGetCapabilities() {
...
-                                COSH, COT, DEGREES, DIV, EXP, FLOOR, GREATEST, LEAST, LN, LOG, MOD, POWER, RADIANS,
-                                RAND, ROUND, SIGN, SIN, SINH, SQRT, TAN, TANH, TRUNC, ASCII, BIT_LENGTH, CHR, CONCAT,
-                                INSTR, LENGTH, LOWER, LPAD, LTRIM, OCTET_LENGTH, REGEXP_REPLACE, REPEAT, REPLACE,
-                                REVERSE, RIGHT, RPAD, RTRIM, SUBSTR, TRANSLATE, TRIM, UNICODE, UNICODECHR, UPPER,
-                                ADD_DAYS, ADD_HOURS, ADD_MINUTES, ADD_MONTHS, ADD_SECONDS, ADD_WEEKS, ADD_YEARS,
-                                SECONDS_BETWEEN, MINUTES_BETWEEN, HOURS_BETWEEN, DAYS_BETWEEN, MONTHS_BETWEEN,
-                                YEARS_BETWEEN, MINUTE, SECOND, DAY, WEEK, MONTH, YEAR, CURRENT_DATE, CURRENT_TIMESTAMP,
-                                DATE_TRUNC, EXTRACT, LOCALTIMESTAMP, POSIX_TIME, TO_CHAR, CASE, HASH_MD5)));
+                                VAR_POP, VAR_POP_DISTINCT, VAR_SAMP, VAR_SAMP_DISTINCT, GROUP_CONCAT)) //
+        );
```
Comment (chiaradiamarcelo) [inline src/test/java/com/exasol/adapter/dialects/postgresql/PostgreSQLSqlDialectTest.java:84]: we should add here a test either for the supported scalar functions or for the non-supported. Ideally for the supported ones as they are part of the public API

File: src/test/java/com/exasol/adapter/dialects/scalarfunction/ScalarFunctionsAbstractIT.java [hunk 3]
```diff
@@ -41,7 +41,7 @@
      */
     private static final Set<ScalarFunctionCapability> EXCLUDES = Set.of(CASE, CAST, FLOAT_DIV, SESSION_PARAMETER, RAND,
```
Comment (chiaradiamarcelo) [inline src/test/java/com/exasol/adapter/dialects/scalarfunction/ScalarFunctionsAbstractIT.java:42]:
```suggestion
    private static final Set<ScalarFunctionCapability> EXCLUDED_SCALAR_FUNCTIONS = Set.of(CASE, CAST, FLOAT_DIV, SESSION_PARAMETER, RAND,
```

File: src/test/java/com/exasol/adapter/dialects/scalarfunction/ScalarFunctionsAbstractIT.java [hunk 4]
```diff
@@ -332,9 +330,11 @@ void afterAll() {
         @ParameterizedTest
         @CsvSource({ "ADD, +, 4", "SUB, -, 0", "MULT, *, 4", "FLOAT_DIV, /, 1" })
         void testSimpleArithmeticFunctions(final ScalarFunctionCapability function, final String operator,
-                final int expectedResult) {
-            if (isTestDisabled(function))
+                final int expectedResult) throws SQLException {
+            if (isTestDisabled(function)) {
```
Comment (chiaradiamarcelo) [inline src/test/java/com/exasol/adapter/dialects/scalarfunction/ScalarFunctionsAbstractIT.java:334]:
```suggestion
            if (isTestDisabledFor(function)) {
```

File: src/test/java/com/exasol/adapter/dialects/scalarfunction/ScalarFunctionsAbstractIT.java [hunk 5]
```diff
@@ -346,14 +346,67 @@ void testSimpleArithmeticFunctions(final ScalarFunctionCapability function, fina
...
+                }
+            });
+        }
+    }
+
+    @Nested
+    @TestInstance(PER_CLASS)
+    @Tag("WithDoubleParameter") // workaround since IDEA can find the test otherwise
+    class WithDoubleParameter {
+
```
Comment (chiaradiamarcelo) [inline src/test/java/com/exasol/adapter/dialects/scalarfunction/ScalarFunctionsAbstractIT.java:382]:
```suggestion
```

File: src/test/java/com/exasol/adapter/dialects/scalarfunction/ScalarFunctionsAbstractIT.java [hunk 6]
```diff
@@ -346,14 +346,67 @@ void testSimpleArithmeticFunctions(final ScalarFunctionCapability function, fina
...
+                try (final ResultSet result = statement.executeQuery(virtualSchemaQuery)) {
+                    assertThat(result, table().row(expectedOutput).matches(TypeMatchMode.NO_JAVA_TYPE_CHECK));
+                }
+            });
+        }
+    }
+
+    @Nested
+    @TestInstance(PER_CLASS)
+    @Tag("WithDoubleParameter") // workaround since IDEA can find the test otherwise
```
Comment (chiaradiamarcelo) [inline src/test/java/com/exasol/adapter/dialects/scalarfunction/ScalarFunctionsAbstractIT.java:380]: I guess you meant
```suggestion
    @Tag("WithDoubleParameter") // workaround since IDEA can't find the test otherwise
```

File: src/test/java/com/exasol/adapter/dialects/scalarfunction/ScalarFunctionsAbstractIT.java [hunk 7]
```diff
@@ -332,9 +330,11 @@ void afterAll() {
         @ParameterizedTest
         @CsvSource({ "ADD, +, 4", "SUB, -, 0", "MULT, *, 4", "FLOAT_DIV, /, 1" })
         void testSimpleArithmeticFunctions(final ScalarFunctionCapability function, final String operator,
-                final int expectedResult) {
-            if (isTestDisabled(function))
+                final int expectedResult) throws SQLException {
+            if (isTestDisabled(function)) {
                 return;
+            }
+            this.valueSetter.setValueOfSingleRow(2);
```
Comment (chiaradiamarcelo) [inline src/test/java/com/exasol/adapter/dialects/scalarfunction/ScalarFunctionsAbstractIT.java:337]: I find a bit confusing and hard to follow the concept the `valueSetter`, as it is an object provided by the `VirtualSchemaTable` to manipulate it.
As I understand, the `valueSetter` is attached to a table, and its responsibility is to restart the table with a single row, meaning basically `truncate` it and then `insert` a single row.

I would prefer to add more meaningful behavior to the `VirtualSchemaTable` by adding this functionality to it, and make the `valueSetter` an implementation detail encapsulated in the `VirtualSchemaTable`.

File: src/test/java/com/exasol/adapter/dialects/scalarfunction/ScalarFunctionsAbstractIT.java [hunk 8]
```diff
@@ -346,14 +346,67 @@ void testSimpleArithmeticFunctions(final ScalarFunctionCapability function, fina
...
+
+        @AfterAll
+        void afterAll() {
+            deleteCreatedTables();
+        }
+
+        @ParameterizedTest
+        @CsvSource({ "0.1, 0", "0.5, 1", "0.9, 1" })
+        void testRound(final double input, final double expectedOutput) throws SQLException {
+            if (isTestDisabled(ROUND))
```
Comment (chiaradiamarcelo) [inline src/test/java/com/exasol/adapter/dialects/scalarfunction/ScalarFunctionsAbstractIT.java:403]: I would prefer a more meaningful name, reflecting the intent. Right now `isTestDisabled` only says that the test is not enabled but it doesn't say why.

Maybe we can change it to something like `isScalarFunctionSupported(ROUND))` ?

File: src/test/java/com/exasol/adapter/dialects/scalarfunction/ScalarFunctionsAbstractIT.java [hunk 9]
```diff
@@ -398,11 +451,11 @@ void testNeg() {
     class WithStringParameter {
         private ScalarFunctionQueryBuilder queryBuilder;
         private String columnName;
-        private StringTestTable.ValueSetter tableRowValueSetter;
+        private VirtualSchemaTestTable.ValueSetter tableRowValueSetter;
```
Comment (chiaradiamarcelo) [inline src/test/java/com/exasol/adapter/dialects/scalarfunction/ScalarFunctionsAbstractIT.java:454]: see my comment about the `valueSetter`

File: src/test/resources/integration/scalarFunctionsParameterCache.yml [hunk 10]
```diff
@@ -1,5 +1,13 @@
-ABS: [ '"FLOATING_POINT"', '"NUMBER"' ]
+ABS: [ '"NUMBER"' ]
```
Comment (chiaradiamarcelo) [inline src/test/resources/integration/scalarFunctionsParameterCache.yml:1]: This document is great, as it describes the test cases for the scalar functions.

Consider to change its name to something more meaningful from the POV of the user, maybe `scalarFunctionsTestCases.yml`

And a small explanation in the `design.md`

File: doc/changes/changes_1.1.0.md [hunk 11]
```diff
@@ -0,0 +1,26 @@
+# Virtual Schema for PostgreSQL 1.1.0, released 2021-??-??
+
+Code name:
+
+## Features / Enhancements
+
+* #8: Add error codes to exceptions
+* #15: Get driver for the integration tests from Maven
+* #13: Added integration test for scalar functions
```
[the remark this answers is not in the document]
Comment (jakobbraun) [reply doc/changes/changes_1.1.0.md:9]: it's the order of fixing. Do you think we should always order them by issue number?

File: src/test/java/com/exasol/adapter/dialects/scalarfunction/ScalarFunctionsAbstractIT.java (hunk 8 above)
Comment (jakobbraun) [reply src/test/java/com/exasol/adapter/dialects/scalarfunction/ScalarFunctionsAbstractIT.java:402]: I don't have more information in that class. It just gives a dialect the possibility to disable certain tests. If it is not supported, one should better remove the capability.

File: src/test/resources/integration/scalarFunctionsParameterCache.yml (hunk 10 above)
Comment (jakobbraun) [reply src/test/resources/integration/scalarFunctionsParameterCache.yml:1]: users should right now not view this file. It is very hard to understand and very misleading.

I would prefer not to describe it in this PR, since it is not Postgresql specific. Maybe when I move the test to vs-common.

Comment (jakobbraun) [reply src/test/resources/integration/scalarFunctionsParameterCache.yml:1]: Especially since the file is not new / changed in this PR ;-)

**Review by chiaradiamarcelo — APPROVED** [review]

File: doc/design.md [hunk 12]
```diff
@@ -0,0 +1,20 @@
+# Design for the PostgreSQL Virtual Schema adapter
+
+## Scalar Functions:
+
+### `MINUTES_BETWEEN`, `HOURSE_BETWEEN` ...
```
Comment (chiaradiamarcelo) [inline doc/design.md:5]:
```suggestion
### `MINUTES_BETWEEN`, `HOURS_BETWEEN` ...
```

File: doc/design.md [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+
+Exasol supports multiple functions that get the diff of a date in a specific unit. For example in Exasol `MINUTES_BETWEEN` of two dates that are two hours apart return 120. For more details check the [Exasol scalar functions manual](https://docs.exasol.com/sql_references/functions/all_functions.htm).
+
+PostgreSQL does not support such functions.
+
+**Design**: Do not support these capabilities.
+
+#### Other considered options:
```
Comment (chiaradiamarcelo) [inline doc/design.md:13]:
```suggestion
#### Considered implementation options:
```

**Comment by chiaradiamarcelo** [comment]
Please don't forget to add a unit test either for the supported scalar functions or for the non-supported. Ideally for the supported ones as they are part of the public API

File: src/test/java/com/exasol/adapter/dialects/scalarfunction/VirtualSchemaTestTable.java [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+    public String getFullyQualifiedName() {
+        return this.fullyQualifiedName;
+    }
+
+    public void setValueOfSingleRow(final T value) throws SQLException {
+        this.valueSetter.setValueOfSingleRow(value);
+    }
+
+    @FunctionalInterface
+    public interface ValueSetter<T> {
```
Comment (chiaradiamarcelo) [inline src/test/java/com/exasol/adapter/dialects/scalarfunction/VirtualSchemaTestTable.java:29]: I think this approach is OK, but we definitely need a better name for this in interface (maybe `VirtualSchemaTableInitializer` or similar), and use a template method to provide the common implementation and the hooks to be extended.

**Review by chiaradiamarcelo — CHANGES_REQUESTED** [review]
Please don't forget to add a unit test either for the supported scalar functions or for the non-supported. Ideally for the supported ones as they are part of the public API

File: doc/design.md [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+
+### `MINUTES_BETWEEN`, `HOURSE_BETWEEN` ...
+
+Exasol supports multiple functions that get the diff of a date in a specific unit. In Exasol `MINUTES_BETWEEN` of two dates that are two hours apart return 120.
+
+PostgreSQL does not have such functions.
+
+**Design**: Do not support these capabilities.
+
+#### Other considered options:
```
[the remark this answers is not in the document]
Comment (jakobbraun) [reply doc/design.md:13]:
```suggestion
#### Other considered solutions:
```

**Review by chiaradiamarcelo — APPROVED** [review]
